src/datatablerow/utils.jl

```diff
@@ -0,0 +1,190 @@
+# Rows grouping.
+# Maps row contents to the indices of all the equal rows.
+# Used by groupby(), join(), nonunique()
+immutable RowGroupDict{T<:AbstractDataTable}
+    dt::T                 # source data frame
+
+    ngroups::Int          # number of groups
+
+    rhashes::Vector{UInt} # row hashes
+    gslots::Vector{Int}   # hashindex -> index of group-representative row
+
+    groups::Vector{Int}   # group index for each row
+    rperm::Vector{Int}    # permutation of row indices that sorts them by groups
+
+    starts::Vector{Int}   # starts of ranges in rperm for each group
+    stops::Vector{Int}    # stops of ranges in rperm for each group
+end
+
+# "kernel" functions for hashrows()
+# adjust row hashes by the hashes of column elements
+function hashrows_col!(h::Vector{UInt}, v::AbstractVector)
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
+        h[i] = isnull(v[i]) ?
+               hash(NULL_MAGIC, h[i]) :
+               hash(v[i], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
+    @inbounds for (i, ref) in enumerate(v.refs)
+        h[i] = hash(CategoricalArrays.index(v.pool)[ref], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractNullableVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractNullableCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
+    @inbounds for (i, ref) in enumerate(v.refs)
+        h[i] = ref == 0 ?
+               hash(NULL_MAGIC, h[i]) :
+               hash(CategoricalArrays.index(v.pool)[ref], h[i])
+    end
+    h
+end
+
+# Calculate hash for each row
+# in an efficient column-wise manner
+function hashrows(dt::AbstractDataTable)
+    res = zeros(UInt, nrow(dt))
+    for col in columns(dt)
+        hashrows_col!(res, col)
+    end
+    return res
+end
+
+# Helper function for RowGroupDict.
+# Returns a tuple:
+# 1) the number of row groups in a data frame
+# 2) vector of row hashes
+# 3) slot array for a hash map, non-zero values are
+#    the indices of the first row in a group
+# Optional group vector is set to the group indices of each row
+function row_group_slots(dt::AbstractDataTable,
+                          groups::Union{Vector{Int}, Void} = nothing)
+    @assert groups === nothing || length(groups) == nrow(dt)
+    rhashes = hashrows(dt)
+    sz = Base._tablesz(length(rhashes))
```

Comment:
One of my intermediate implementations was using `Dict{DataFramesRow, ...}`, but the current one has more efficient memory access patterns when generating hashes (by column instead of by row). And it doesn't allocate memory for Dict values (the values would be the sets of row indices from the same group), because this information could be efficiently retrieved from `RowGroupDict` (`rperm`, `starts` etc -- these are just vectors of integers).